File: src/main/java/com/DavidM1A2/AfraidOfTheDark/common/block/core/AOTDSapling.java
/*
 * Author: David Slovikosky
 * Mod: Afraid of the Dark
 * Ideas and Textures: Michael Albertson
 */
package com.DavidM1A2.AfraidOfTheDark.common.block.core;

import java.util.Random;

import com.DavidM1A2.AfraidOfTheDark.common.reference.Reference;

import net.minecraft.block.BlockBush;
import net.minecraft.block.IGrowable;
import net.minecraft.block.SoundType;
import net.minecraft.block.properties.IProperty;
import net.minecraft.block.properties.PropertyInteger;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;

public abstract class AOTDSapling extends BlockBush implements IGrowable
{
	public static final PropertyInteger STAGE_PROP = PropertyInteger.create("stage", 0, 1);
	protected static final AxisAlignedBB SAPLING_AABB = new AxisAlignedBB(0.09999999403953552D, 0.0D, 0.09999999403953552D, 0.8999999761581421D, 0.800000011920929D, 0.8999999761581421D);

	public AOTDSapling()
	{
		super();
		this.setHardness(0.0F);
		this.setSoundType(SoundType.PLANT);
		this.setDefaultState(this.blockState.getBaseState().withProperty(STAGE_PROP, Integer.valueOf(0)));
		this.setCreativeTab(Reference.AFRAID_OF_THE_DARK);
	}

	@Override
	public AxisAlignedBB getBoundingBox(IBlockState state, IBlockAccess source, net.minecraft.util.math.BlockPos pos)
	{
		return SAPLING_AABB;
	}

	@Override
	public void updateTick(World world, BlockPos blockPos, IBlockState iBlockState, Random random)
	{
		if (!world.isRemote)
		{
			super.updateTick(world, blockPos, iBlockState, random);

			if (world.getLightFromNeighbors(blockPos.up()) >= 9 && random.nextInt(7) == 0)
			{
				this.grow(world, blockPos, iBlockState, random);
			}
		}
	}

	public void grow(World world, BlockPos blockPos, IBlockState iBlockState, Random random)
	{
		if (((Integer) iBlockState.getValue(STAGE_PROP)).intValue() == 0)
		{
			world.setBlockState(blockPos, iBlockState.cycleProperty(STAGE_PROP), 4);
		}
		else
		{
			this.causeTreeToGrow(world, blockPos, iBlockState, random);
		}
	}

	public abstract void causeTreeToGrow(World world, BlockPos blockPos, IBlockState iBlockState, Random random);

	/**
	 * Get the damage value that this Block should drop
	 */
	@Override
	public int damageDropped(IBlockState state)
	{
		return 0;
	}

	@Override
	public boolean canUseBonemeal(World world, Random random, BlockPos blockPos, IBlockState iBlockState)
	{
		return world.rand.nextFloat() < 0.45D;
	}

	@Override
	public void grow(World world, Random random, BlockPos blockPos, IBlockState iBlockState)
	{
		this.grow(world, blockPos, iBlockState, random);
	}

	/**
	 * Convert the given metadata into a BlockState for this Block
	 */
	@Override
	public IBlockState getStateFromMeta(int meta)
	{
		return this.getDefaultState().withProperty(STAGE_PROP, Integer.valueOf((meta & 8) >> 3));
	}

	/**
	 * Convert the BlockState into the correct metadata value
	 */
	@Override
	public int getMetaFromState(IBlockState state)
	{
		byte b0 = 0;
		int i = b0;
		i |= ((Integer) state.getValue(STAGE_PROP)).intValue() << 3;
		return i;
	}

	@Override
	protected BlockStateContainer createBlockState()
	{
		return new BlockStateContainer(this, new IProperty[]
		{ STAGE_PROP });
	}

	@Override
	public String getUnlocalizedName()
	{
		return String.format("tile.%s%s", Reference.MOD_ID.toLowerCase() + ":", this.getUnwrappedUnlocalizedName(super.getUnlocalizedName()));
		// Format for a block is: tile.modid:blockname.name
	}

	// Get the unlocalized name
	protected String getUnwrappedUnlocalizedName(final String unlocalizedName)
	{
		return unlocalizedName.substring(unlocalizedName.indexOf(".") + 1);
	}
}

File: src/main/java/com/davidm1a2/afraidofthedark/common/block/core/AOTDSapling.kt
package com.davidm1a2.afraidofthedark.common.block.core

import com.davidm1a2.afraidofthedark.common.constants.Constants
import net.minecraft.block.BlockBush
import net.minecraft.block.IGrowable
import net.minecraft.block.SoundType
import net.minecraft.block.properties.PropertyInteger
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import java.util.*

/**
 * Class representing the base for a sapling created in afraid of the dark
 *
 * @constructor sets up default sapling properties
 * @param baseName The name of the block to register
 */
abstract class AOTDSapling(baseName: String) : BlockBush(), IGrowable
{
    init
    {
        unlocalizedName = "${Constants.MOD_ID}:$baseName"
        this.setRegistryName("${Constants.MOD_ID}:$baseName")
        setHardness(0.0f)
        setResistance(0.0f)
        this.soundType = SoundType.PLANT
        setLightLevel(0.0f)
        setLightOpacity(0)
        setCreativeTab(Constants.AOTD_CREATIVE_TAB)
    }

    /**
     * Saplings have a special bounding box independent of anything else, so just reutrn the constant bounding box
     *
     * @param state  The state of this sapling
     * @param source Tests if the block can access light
     * @param pos    The position of this sapling
     * @return Sapling bounding box
     */
    override fun getBoundingBox(state: IBlockState, source: IBlockAccess, pos: BlockPos): AxisAlignedBB
    {
        return SAPLING_AABB
    }

    /**
     * Called randomly on this block every few ticks
     *
     * @param world  The world this sapling is ticking in
     * @param pos    The position of this sapling
     * @param state  The state of this sapling
     * @param random The current random object
     */
    override fun updateTick(world: World, pos: BlockPos, state: IBlockState, random: Random)
    {
        // Server side only
        if (!world.isRemote)
        {
            // Perform any original functionality
            super.updateTick(world, pos, state, random)

            // If the area of the world is not loaded, return
            if (!world.isAreaLoaded(pos, 1))
            {
                return
            }

            // If we have enough light and our 1 in 7 chance succeeds, advance our sapling growth
            if (world.getLightFromNeighbors(pos.up()) >= 9 && random.nextInt(7) == 0)
            {
                advanceGrowth(world, pos, state, random)
            }
        }
    }

    /**
     * Called to advance the growth of our tree. If it has grown enough, create a tree
     *
     * @param world  The world the tree is growing in
     * @param pos    The position of the sapling
     * @param state  The current state of the sapling
     * @param random The random object associated with this sapling
     */
    private fun advanceGrowth(world: World, pos: BlockPos, state: IBlockState, random: Random)
    {
        // If the sapling has never grown before, advance it's growth to the next stage. When advance growth is called again the sapling will grow
        if (state.getValue(STAGE) == 0)
        {
            world.setBlockState(pos, state.cycleProperty(STAGE), 4)
        }
        else
        {
            causeTreeToGrow(world, pos, state, random)
        }
    }

    /**
     * Called to make the tree grow. The algorithm is tree type dependent!
     *
     * @param world  The world the sapling is growing in
     * @param pos    The position of the sapling
     * @param state  The current state of the sapling
     * @param random The random object to grow the tree with
     */
    abstract fun causeTreeToGrow(world: World, pos: BlockPos, state: IBlockState, random: Random)

    /**
     * Test if the sapling can grow or not, will always return true
     *
     * @param world    The world the sapling is growing in
     * @param pos      The position of the sapling
     * @param state    The current state of the sapling
     * @param isClient If the function is being called client or server side
     * @return True if the sapling can grow, false otherwise
     */
    override fun canGrow(world: World, pos: BlockPos, state: IBlockState, isClient: Boolean): Boolean
    {
        return true
    }

    /**
     * Tests if the sapling is bone-meal able. True 55% of the time randomly
     *
     * @param world  The world the sapling is growing in
     * @param random The random object to grow the tree with
     * @param pos    The position of the sapling
     * @param state  The current state of the sapling
     * @return True if the bonemeal was successful, false otherwise
     */
    override fun canUseBonemeal(world: World, random: Random, pos: BlockPos, state: IBlockState): Boolean
    {
        return world.rand.nextFloat() < 0.45f
    }

    /**
     * Called to force this sapling to grow, just calls advance growth
     *
     * @param world  The world the sapling is growing in
     * @param random The random object to grow the tree with
     * @param pos    The position of the sapling
     * @param state  The current state of the sapling
     */
    override fun grow(world: World, random: Random, pos: BlockPos, state: IBlockState)
    {
        advanceGrowth(world, pos, state, random)
    }

    /**
     * Converts from metadata integer to block state
     *
     * @param meta The metadata value to convert
     * @return The blockstate to convert to
     */
    override fun getStateFromMeta(meta: Int): IBlockState
    {
        return this.defaultState.withProperty(STAGE, meta)
    }

    /**
     * Converts from state object to metadata integer
     *
     * @param state The block state to convert into integer metadata
     * @return The integer metadata representing this state
     */
    override fun getMetaFromState(state: IBlockState): Int
    {
        return state.getValue(STAGE)
    }

    /**
     * Makes sure to give this block a state
     *
     * @return The block state container with all properties this sapling defines
     */
    override fun createBlockState(): BlockStateContainer
    {
        return BlockStateContainer(this, STAGE)
    }

    companion object
    {
        // What stage of growth the sapling is in
        private val STAGE = PropertyInteger.create("stage", 0, 1)
        // The bounding box of the sapling
        private val SAPLING_AABB = AxisAlignedBB(0.09999999403953552, 0.0, 0.09999999403953552, 0.8999999761581421, 0.800000011920929, 0.8999999761581421)
    }
}